reassembly/tcpassembly.go

```diff
@@ -0,0 +1,1227 @@
+type LivePacket struct {
+	bytes []byte
+	start bool
+	end   bool
+	ci    gopacket.CaptureInfo
+	ac    AssemblerContext
+	seq   Sequence
+}
+
+func (lp *LivePacket) Bytes() []byte {
+	return lp.bytes
+}
+func (lp *LivePacket) CaptureInfo() gopacket.CaptureInfo {
+	return lp.ci
+}
+func (lp *LivePacket) AssemblerContext() AssemblerContext {
+	return lp.ac
+}
+func (lp *LivePacket) Length() int {
+	return len(lp.bytes)
+}
+func (lp *LivePacket) Start() bool {
+	return lp.start
+}
+func (lp *LivePacket) End() bool {
+	return lp.end
+}
+func (lp *LivePacket) Seq() Sequence {
+	return lp.seq
+}
+func (lp *LivePacket) IsPacket() bool {
+	return true
+}
+
+// Creates a page (or set of pages) from a TCP packet: returns the first and last
+// page in its doubly-linked list of new pages.
+func (lp *LivePacket) ConvertToPages(pc *pageCache, skip int, ac AssemblerContext) (*page, *page, int) {
+	ts := lp.ci.Timestamp
+	first := pc.next(ts)
+	current := first
+	first.isPacket = true
+	numPages := 1
+	seq, bytes := lp.seq.Add(-skip), lp.bytes[skip:]
+	for {
+		length := min(len(bytes), pageBytes)
+		current.bytes = current.buf[:length]
+		copy(current.bytes, bytes)
+		current.seq = seq
+		current.ac = ac
+		bytes = bytes[length:]
+		if len(bytes) == 0 {
+			current.end = lp.End()
+			break
+		}
+		seq = seq.Add(length)
+		current.next = pc.next(ts)
+		current.next.prev = current
+		current = current.next
+		numPages++
+	}
+	return first, current, numPages
+}
+func (lp *LivePacket) estimateNumberOfPages() int {
+	return (len(lp.bytes) + pageBytes + 1) / pageBytes
+}
+
+func (lp *LivePacket) Release(*pageCache) int {
+	return 0
+}
+
+// Stream is implemented by the caller to handle incoming reassembled
+// TCP data.  Callers create a StreamFactory, then StreamPool uses
+// it to create a new Stream for every TCP stream.
+//
+// assembly will, in order:
+//    1) Create the stream via StreamFactory.New
+//    2) Call ReassembledSG 0 or more times, passing in reassembled TCP data in order
+//    3) Call ReassemblyComplete one time, after which the stream is dereferenced by assembly.
+type Stream interface {
+	// Tell whether the tcp packet should be accepted, start could be modified to force a start even if no SYN have been seen
+	Accept(tcp *layers.TCP, ci gopacket.CaptureInfo, dir TcpFlowDirection, ackSeq Sequence, start *bool, ac AssemblerContext) bool
+	// ReassembledSG is called zero or more times.
+	// Reassembly objects are reused after each Reassembled call,
+	// so it's important to copy anything you need out of them
+	// or to not call Release() to keep those bytes
+	ReassembledSG(sg ScatterGather, ac AssemblerContext)
+	// ReassemblyComplete is called when assembly decides there is
+	// no more data for this Stream, either because a FIN or RST packet
+	// was seen, or because the stream has timed out without any new
+	// packet data (due to a call to FlushCloseOlderThan).
+	// It returns true if connection should be removed from the pool
```

Comment:
Yes, in that case we keep the stream around to allow tracking the end of the connection (the TCP LAST-ACK state).